### ISSUE TITLE
Decide on the role of `:` in indentation syntax

### ISSUE THREAD

**Issue by odersky** [lampepfl/dotty#7136]
The meaning of `:` for significant indentation is more contentious than other aspects. We should come up with a crisp and intuitive definition where `:` is allowed.

One way I'd like to frame significant indentation in Scala 3, is that _braces are optional_, analogously to how semicolons are optional. But what does that mean? If we disregard `:` it means:

> At all points where code in braces `{...}` is _allowed_, and _some_ subordinate code is _required_, an indented code block is treated as if it was in braces.

The second condition is important, since otherwise any deviation from straight indentation would be significant, which would be dangerous and a strain on the eyes. But with that condition it's straightforward: If some subordinate code is required, either that code follows on the same line, or it is on a following line, in which case it should be indented. Braces are then optional, they are not needed to decide code structure. In the following I assume this part as given.

So that leaves the places where some code is not required but we still like to insert braces. There are two reasonable use cases for this:

 1. Between the extends clause of a class, object, or similar and its (optional) template definitions in braces. 
 2. In a partial application such as
    ```scala
    xs.foreach:
      ...
    ```
    The motivation for this second case is to make uses of library-defined operations close to
    native syntax. If native syntax allows to drop braces, there should be a way for library-defined
    syntax to do the same.

### Possible schemes

The current scheme relies on colons at end of lines that can be inserted in a large number of situations. There are several possible alternative approaches I can see:

1. Don't do it. Insist on braces for (1) and (2).

2. Split (1) and (2). Make indentation significant after class, object, etc headers without requiring a semicolon. This has the problem that it is not immediately clear whether we define an empty template or not. E.g. in
    ```scala
    object X extends Z

     // ...
     // ...

     object Y
    ```
   it is hard to see whether the second `object` is on the same level as the first or subordinate. But semantically it makes a big difference. So a system like that would be fragile. By contrast, a mandatory `:` would make it clear. Then the version above would be two objects on the same level and to get a subordinate member object Y you'd write instead:
    ```scala
    object X extends Z:

     // ...
     // ...

     object Y
    ```
   So I actually quite like the colon in this role.

3. Split (1) and (2) but require another keyword to start template definitions after class, object, etc headers. @eed3si9n suggested `where`. It's a possibility, but again I do like `:` at this point. It reads better IMO (and `where` might be a useful keyword to have elsewhere, e.g. in a future Scala with predicate refinement types).

4. Keep `:` (or whatever) to start templates but introduce a general "parens-killing" operator such as `$` in Haskell with a mandatory RHS argument. If that occurred at the end of a line, braces would be optional according to our ground rules. 

   I fear that a general operator like that would lead to code that was hard to read for non-experts. I personally find `$` did a lot more harm than good to the readability of Haskell code.

5. Restrict the role of `:` to the two use cases above. I.e. allow a colon if

    - a template body is expected, or
    - in a partial application, after an identifier, closing bracket `]` or closing parenthesis `)`.

   This way, we would avoid the [confusing line noise](https://github.com/lampepfl/dotty/pull/7083#issuecomment-526282840) that can happen if we allow `:` more freely.

My personal tendency would be to go for that last option.

**Comment by LPTK** [comment]
Regardless of whether we find a solution for (2), I would go for requiring braces for (1) — object, class, and trait definitions — which is what I proposed [in the other thread](https://github.com/lampepfl/dotty/pull/7083#issuecomment-524550858).

But let me repeat and strengthen the rationale here:

 * It provides a minimum amount of guidance while reading big nested definitions.

    I already find it _really difficult_ to read sources [like this](https://github.com/scala/scala/blob/42952e61ae4f23be573250c93da4a87db2fa62b4/src/reflect/scala/reflect/internal/Symbols.scala) or [this](https://github.com/scala/scala/blob/42952e61ae4f23be573250c93da4a87db2fa62b4/src/reflect/scala/reflect/internal/Types.scala), where there are **several levels of nested objects/classes with huge bodies**. It's really challenging to see what begins and ends where — just imagine what it would look like without braces. Currently, using an IDE helps, as I can place the cursor on an opening _or_ closing brace and see the whole corresponding scope highlighted/delineated. That won't be possible without braces.

    And as others have pointed out, this kind of code does come up in realistic situations (my examples are from the scala/scala repo after all). If we make braces around classes/objects optional, we'll encourage people not to use them, and we'll invariably end up with unreadable messes when files naturally grow in size over time.

 * It _makes sense conceptually_, if we consider that curly braces are used specifically to denote object scopes.

 * I would also argue that object scopes should be explicitly delineated because they are _much more significant_ than expression scopes. Indeed, in an object scope, you get a bunch of implicitly imported definitions, you get a new meaning for `this`, etc.

**Comment by odersky** [comment]
@LPTK I believe that end markers are actually a far superior way to delineate class scopes. And having  both braces and end markers would look weird.

**Comment by JanBessai** [comment]
Not sure if this is the correct thread, but:
How does indentation sensitivity mix with triple quoted multi-line strings? Does the end-quote have to be indented? If so: how do you write down a line break + spaces before the end quote? Indent further?

**Comment by sjrd** [comment]
At the risk of having to my own devil's advocate in a week or so ... what about using `with` instead of `:` for the two use cases above.

For templates:
```scala
class Foo extends Bar with SomeTrait with
  def x: Int = 42
```
The `with` there does not shock me. It is even quite easy to interpret it as "braces as optional" if we also allow
```scala
class Foo extends Bar with SomeTrait with {
  def x: Int = 42
}
```
That syntax makes sense to me, as the contents of the block are added as members of `Foo` just as much as the members of `SomeTrait`. I don't know, it kind of makes sense to me.

For that, the lexer would introduce an indent if `with` is at the end of the line, the next line is indented *and* starts by anything but an identifier.

I would actually prefer to simply *not* have anything to open a template, though. However this might require the parser to feed into the lexer to implement, which is not ideal. Or maybe we can make it work by simply counting opening and closing brackets? That would be nice.

For method calls:
```scala
xs.foreach with
  x =>
    println(x)

val y = xs.foldLeft(0) with
  (prev, x) =>
    prev + x

val z = optInt.fold with
  println("nope")
with
  x => println(x)
```
This syntax had actually been proposed in the original "new implicits" proposals, for implicit parameter lists. So it must have appealed at some point. Except here it is used to pass normal arguments, when they are blocks (by-name params or lambdas).

The advantage over `:` is that, well, it's not `:`. There have been several concerns about the multiple problems that `:` exposes, not the least being the overloading wrt. type ascriptions.

`with` also suffers from overloading in this case, but perhaps it is less annoying than `:` because it would only have two meanings that are very well visually separated: `with` in a class header is composition; `with` in an expression is a block argument.

I don't really like what I'm proposing, but I dislike it a lot less than `:`.

**Comment by jducoeur** [comment]
> I don't really like what I'm proposing, but I dislike it a lot less than :.

Strong agreement.  I think `:` is terrible -- not only is it badly overloaded, it's simply too visually subtle.  I'm concerned that it will lead to bugs due to people just missing it.  I dislike this whole approach (IMO it crosses the line into designing a different language, not enhancing Scala, and enormously increases the risk of splitting the community) but probably this aspect most of all...

**Comment by eed3si9n** [comment]
In [#7083](https://github.com/lampepfl/dotty/pull/7083#issuecomment-524392428) I wrote:

> Since colon means type annotation (ascription) `x: A` already, reusing this to mean "begin bill of material" or "begin block" seems odd to me too.
> 
> F# uses `=`, and Haskell uses [`where`](https://www.haskell.org/onlinereport/haskell2010/haskellch5.html):

```haskell
module Main where  
  import A  
  import B  
  main = A.f >> B.f
```

## (2) In a partial application such as xs.foreach:

I think `with` is an improvement over `:`, but it does suffer from overloading. Could we borrow `<|` from F# here?

```scala
xs.foreach <| x =>
  println(x)

ys.foreach <| y => println(y)

val y = xs.foldLeft(0) <| (prev, x) =>
  prev + x
```

If you squint, it looks like begin `{`. Note here that the indentation doesn't start until `\n` so you can write `x =>` on the same line.

## pattern match

Pattern matching is an odd one because `<|` is the opposite direction.

```scala
val kind = ch match
  case ' '  => "space"
  case '\t' => "tab"
  case _    => s"'$ch'-character"
```

You basically want `|>` but that might be too confusing:

```scala
val kind = ch |>
  case ' '  => "space"
  case '\t' => "tab"
  case _    => s"'$ch'-character"
```

I think we should just allow `match` to be a special indentation introducer.

## (1) template definitions

What's interesting about template is that it is both a list of members and it's also the body of the constructor. So in that sense it might make sense to keep it the same syntax as block introducer.

```scala
class Contact(name: String) extends Bar with SomeTrait <|
  def x: Int = 42

  DB.append(name)

object Contact <|
  def apply: Contact = new Contact("")
```

I understand that `=` is not exactly right, but visually it looks less intimidating I think

```scala
class Contact(name: String) extends Bar with SomeTrait =
  def x: Int = 42

  DB.append(name)

object Contact =
  def apply: Contact = new Contact("")
```

> I don't really like what I'm proposing, but I dislike it a lot less than `:`.

Ditto.

**Comment by lihaoyi** [comment]
I propose re-using existing keywords wherever possible. The whole point of this exercise is to make the syntax lightweight: having long keywords like `where` defeats the purpose entirely. if necessary, we could commandeer `do` as a short-enough keyword to delimit blocks in the case of ambiguity

Re-using existing keywords can get us surprisingly far. Consider [Scalite](https://github.com/lihaoyi/scalite):

```scala
package scalite.tutorial                                    package scalite.tutorial

class Point(xc: Int, yc: Int)                               class Point(xc: Int, yc: Int) {
    var x: Int = xc                                           var x: Int = xc
    var y: Int = yc                                           var y: Int = yc
    def move(dx: Int, dy: Int) =                              def move(dx: Int, dy: Int) = {
        x = x + dx                                              x = x + dx
        y = y + dy                                              y = y + dy
                                                              }
    override def toString() =                                 override def toString() = {
        "(" + x + ", " + y + ")"                                "(" + x + ", " + y + ")"
                                                              }
                                                            }
object Run                                                  object Run {
    def apply() =                                             def apply() = {
        val pt = new Point(1, 2)                                val pt = new Point(1, 2)
        println(pt)                                             println(pt)
        pt.move(10, 10)                                         pt.move(10, 10)
        pt.x                                                    pt.x
                                                              }
                                                            }
```

```scala
var x = 0                                                   var x = 0
for(i <- 0 until 10)                                        for(i <- 0 until 10) {
    val j = i * 2                                             val j = i * 2
    val k = j + 1                                             val k = j + 1
    x += k                                                    x += k
                                                            }
val list =                                                  val list = {
    for(i <- 0 to x) yield                                    for(i <- 0 to x) yield {
        val j = i + 1                                           val j = i + 1
        i * j                                                   i * j
                                                              }
                                                            }
list.max                                                    list.max
// 10100                                                    // 10100
```
```scala
val all = for                                               val all = for {
    x <- 0 to 10                                              x <- 0 to 10
    y <- 0 to 10                                              y <- 0 to 10
    if x + y == 10                                            if x + y == 10
yield                                                       } yield {
    val z = x * y                                             val z = x * y
    z                                                         z
                                                            }
all.max                                                     all.max
// 25                                                       // 25
```

I think this looks superior to using any delimiter. Note that the above *already works in Scala 2.11/12/13*, and has worked for 5 years now.

We have to use the same whitespace rules for both `def`/`var`/`val`s and `class`/`trait`/`object`. Having mixed rules for what things are "whitespace compatible" and what things aren't is a mess, and would look terribly ugly, especially in Scala where (unlike Java) having nested inner `class`/`object`s at the same level as your `def`/`var`/`val`s is common, and having top-level `def`/`var`/`val`s at the same level as your top-level `class`/`object`s is going to become common as well. If we want classes/objects to be "special", that ship sailed long ago.

One subtlety that we have to take care of is providing for higher-order methods. Code like:

```scala
val foo = bar
  .map{x => 
    val y = x + 1
    y + 1
  }
  .foreach{ x =>
    val y = x + 1
    println(y)
  }
```

is extremely common with Scala's method-chaining conventions, and I'd want to be able to provide a whitespace-compatible syntax:

```scala
val foo = bar
  .map x => 
    val y = x + 1
    y + 1
  .foreach x =>
    val y = x + 1
    println(y)
```

In Scalite I commandeered the `do` keyword for this purpose:

```scala
val xs = 0 until 10                                         val xs = 0 until 10
val ys = xs.map do                                          val ys = xs.map{
    x => x + 1                                                x => x + 1
                                                            }
ys.sum                                                      ys.sum
// 55                                                       // 55

val zs = xs.map do                                          val zs = xs.map{
case 1 => 1                                                   case 1 => 1
case 2 => 2                                                   case 2 => 2
case x if x % 2 == 0 => x + 1                                 case x if x % 2 == 0 => x + 1
case x if x % 2 != 0 => x - 1                                 case x if x % 2 != 0 => x - 1
                                                            }
zs.sum                                                      zs.sum
// 45                                                       // 45
```
```scala
val ws = xs.map do x =>                                     val ws = xs.map { x =>
    val x1 = x + 1                                            val x1 = x + 1
    x1 * x1                                                   x1 * x1
                                                            }
ws.sum                                                      ws.sum
// 385                                                      // 385
```

**Comment by odersky** [comment]
I want to comment on sub-part (2), i.e. what to use for starting an indented argument. I believe that no single keyword would work well in that role. 

`with` was bordeline acceptable as an implicit argument, since it evokes pairing with a context. But for plain arguments it looks wrong at least as often as it looks OK. For instance:
```scala
print with 
  val x = f(y)
  x * x
```
I believe `with` would only work well if the real function application is to something else and the part following `with` is in some way an accessory to that. E.g. in `xs.map with f` it works since the mapped argument is `xs` and `f` is an accessory. But the case of `print` above shows that we cannot generalize that to all applications.

`do` evokes side effects. It is used now in that role specifically in while or for loops. E.g.
```scala
for 
  x <- xs
  y <- ys
do
  println(x + y)
```
I believe for pure functions, `do` is out of place. E.g.
```scala
transpose do
  val a: Matrix = ...
  val b: Matrix = ...
  a * b
```
In fact, the pattern of function application is so general and multi-faceted that no single keyword can do it justice.  That's why Dijkstra uses infix point (well, that's taken already in Scala!) and Haskell uses `$`. But colon does work brilliantly in this role! Evidence #1 is Python, where it is perceived to be very natural. Yes, I know Python uses it everywhere instead of just for this purpose, but still... Evidence #2 is common language. It's very natural to write something like

>  For making a cake:
>      Preheat oven
>      Mix flours, eggs and milk
>      Bake for one hour

The `:` specifically introduces a list of statements that's subordinate to a prefix clause. That's [completely grammatical](https://www.thepunctuationguide.com/colon.html). So, I believe strongly that `:` is the best operator for this. 

As to possible ambiguity with type ascription: I really don't think that's a problem. `:` at the end of line and `:` used infix are visually quite distinct. The only caveat is that we should not let one follow the other. I.e.
```scala
  def f(): T:
     return foo
```
would be awkward. But ever since procedure syntax was dropped, Scala does not have syntax where this pattern could occur. 

The only possible ambiguity is in a type ascription of an expression spanning multiple lines like this:
```scala
  someLongExpression:
    someLongType
```
But I have not seen code like this in the wild, and in fact our code base including all tools, all tests, and community build does not have a single instance where this pattern occurs. Why does it not occur?
I guess if you have long expressions and types to combine you realize that you are probably better off factoring stuff into a `val`:
```scala
  val someId: someLongType = 
    someLongExpression
  someId
```
And, if you really need to write a multi-line ascription, you can aways do 
```scala
  someLongExpression
    : someLongType
```
which in fact reads much better. So, in summary, any ambiguity would be extremely rare, and is easily avoided. The other evidence why ambiguities are not a concern is again Python. Python does use `:` for both roles and the Python community is not known to be sloppy with syntax. Interestingly Python does _not_ use `:` to indicate a function return type, since that would let them run into precisely the awkwardness I referred to earlier. It uses `->` instead. But since Scala uses `=` instead of `:` to start a function body it  does not run into that problem.

One downside of `:` is that it only works at the end of lines. So
```scala
  xs.map: x =>
    val y = f(x)
    g(y)
```
does not work. You have to format it instead as
```scala
  xs.map: 
    x =>
    val y = f(x)
    g(y)
```
I think this is not so bad. In real code the `{ x =>` part is often quite far to the right because the expression preceding it is long. This makes it hard to see the bound name. The vertical syntax makes it much clearer what is defined.

**Comment by lihaoyi** [comment]
> I think this is not so bad. In real code the { x => part is often quite far to the right because the expression preceding it is long. This makes it hard to see the bound name. The vertical syntax makes it much clearer what is defined.

To me this formatting is he deal breaker, much more than the ambiguity around type ascriptions. I work with a lot of very-much-real code formatted exactly as you describe, and it reads excellently. If the LHS is long, the .map goes on a new line.

I have seen no code at all formatted similar to how you propose, and subjectively it looks awful. If it really looked better, people would already be formatting their lambdas like that right now, and they’re not.

**Comment by odersky** [comment]
@lihaoyi 

> I have seen no code at all formatted similar to how you propose, and subjectively it looks awful. If it really looked better, people would already be formatting their lambdas like that right now, and they’re not.

Fair point. 

Here's a crazy idea for this pattern: use `case`. I.e.
```scala
xs.map case x =>
    val y = f(x)
    g(y)

xs.collect case Some(n) => n

xs.foreach case i => println(s"next: $i")

try body
catch case ex: IOException => 
  log(ex)
  throw ex
```
Points in favor: 

 - `case` is semantically a correct option here.
 - we can combine this with a pattern match.
 - indentation works out of the box, but now it's prompted by the `=>` at the EOL.

To make this work we'd have to add one production to `Expr1`. It's the last line below:

Expr1 ::= ...
             |  Expr2 ‘match’ ‘{’ CaseClauses ‘}’
             |  Expr2 ‘case’ Pattern [Guard] ‘=>’ Expr

```
WDYT?

**Comment by odersky** [comment]
[Aside: You may or may not have noted that I "ate my own dog food" in the comments above: every single indented section was introduced with `:`. Readers should judge for themselves whether this is natural or not.]

**Comment by lihaoyi** [comment]
`case` looks ok-ish to me, but that seems to introduce an even worse ambiguity: that between partial functions and total functions! This is something we had already made efforts to disambiguate in Dotty (e.g. requiring `case` for partial functions for `for`-comprehensions), so introducing ambiguity in a different area seems like a step backwards.

Honestly, I think we should just use `do`. The English meaning is almost irrelevant: `with`, `class`, `object`, `for`, etc. in Scala already mean vastly different things in English and Scala, and that's been mostly OK. Having `do` mean something specific in Scala isn't going to be the end of the world, especially since it's going to be entirely unambiguous and consistent: `do` will have no other meanings once do-while loops are out. I think it would definitely be preferable to overloading `with` or `case` or some other keyword which have very specific, existing Scala meanings.

I appreciate the desire to use `:`, but function-literals-with-arguments is a real sticking point. It does not make sense to have a syntax where passing blocks is easy, passing by-name-params (basically 0-arg functions) is easy, and then passing 1-arg or 2-arg or n-arg functions has some totally different/worse syntax.

Python can get by with (1) not having multiline lambdas (2) a different `-> foo` syntax for annotating return types and (3) a dedicated syntax for context managers. Scala has none of these things, and passing multi-line n-arg function literals to higher order functions is the way people work. We have to make sure it looks pretty and first class

**Comment by odersky** [comment]
I have strong objections to `do`. `do` universally means side effect, in natural language as well as in all programming languages I know. (`do` comprehensions in Haskell model side effects via monads). So I believe we cannot simply change its meaning, in particular in a language like Scala which is predominantly functional but still allows side effects.

`case` is often used with partial functions but not always. A counter example is:
```scala
case class Point(x: Double, y: Double)

points.map {
  case Point(x, y) => ...
}
```
So, `case` is also used for destructuring in total functions. A simple variable pattern is a special case of that (and it's obvious at a glance that that's what it is)

**Comment by lihaoyi** [comment]
I'd argue that the vast majority of `do`s in any language are not in Haskell (or Java/C/C++/Javascript/etc., where do-while loops are uncommon) but in Ruby, which uses `do` exactly to delimit lambdas which can have a return value:

```ruby
lihaoyi ~$ cat foo.rb
def my_map(array)
  new_array = []

  for element in array
    new_array.push yield element
  end

  new_array
end

result = my_map([1, 2, 3]) do |number|
  number * 2
end

puts result.to_s
lihaoyi ~$ ruby foo.rb
[2, 4, 6]

**Comment by Ichoran** [comment]
The niche a language occupies is a highly relevant consideration when choosing its syntax.  Unless we want Scala to stop trying to occupy the "ultra-powerful type system" niche, where keeps pace with Haskell, `do` is a bad idea, even though the keyword is available.

I also think precedent is very important when deciding whether a language syntax is a good idea or not.  I fully agree that `:` is completely natural linguistically--it's frequently used for introducing a following bunch of stuff.  However, we're all really well-trained to see `:` as type ascription (either an actual type or a typeclass), and that makes me hesitant--even aside from genuine parse ambiguities--to repurpose it as a begin-block symbol.

With regards to `case`--well, if we must, okay, but like Li Haoyi I find it a wart in general.  Rust does fine without it, and I wish we could too; or at least, I wish we could restrict it to only partial functions.  Total functions, even with multiple match blocks, would be nicer without `case`.

Finally, I don't think my objections to having a bunch of different block-introductions have been adequately addressed.  If we do pick something, I think it should be universal, even if it admits weird stuff.  I would rather allow

```
foo:
  3
: 
  4
```

and insist on

```
if p then:
  foo
else:
  bar
```

than have to try to intuit which things require `:` and which do not.  In addition to the keyword confusions I mentioned before, it also makes a sharper distinction between builtin language features and added syntax.  One of the most beautiful features of Scala is that you can define your own libraries that feel perfectly seamless, as if they're a built-in part of the language.  Having special rules for where `:` is used and where not would break this.

**Comment by lihaoyi-databricks** [comment]
> I think it should be universal, even if it admits weird stuff. I would rather allow
> ```
> foo:
>   3
> : 
>   4
> ```

I think this would look pretty reasonable with a keyword like `do`:

```
foo do
  ???
do
  ???
```


Each indented `do` block is equivalent to a pair of braces. Although I wonder if it's possible to make the lexer/parser smart enough to omit the first `do`?

```
foo
  ???
do
  ???
```

This would look identical to a while-loop:

```
while
  ???
do
  ???
```

Or if-else

```
if(???)
  ???
else
  ???
```

That would essentially put user-land code syntactically on even footing with the builtin constructs, which seems like exactly what @Ichoran wants

**Comment by lihaoyi-databricks** [comment]
> Unless we want Scala to stop trying to occupy the "ultra-powerful type system" niche, where keeps pace with Haskell

This seems like a reasonable thing to me. I've been using Scala for almost a decade and ~never written any Haskell-ish code, and Scala is its own language with its own style and conventions.

I'm pretty sure this whole whitespace experiment was to try and emulate the approachability and widespread appeal of Python. We aren't trying to attract Haskellite's to Scala. That would make "keeping pace with Haskell" a non-goal altogether (though it still is unclear to me how the spelling of keyword affects the power of the type system)

**Comment by odersky** [comment]
@lihaoyi 

After thinking a bit more about it I believe the `case` idea is indeed a bit crazy, since it risks overloading meanings, as you say. But `do` is not a workable choice either, for the reasons I stated. I believe Ruby's precendent is not a real counter-example: Ruby (and Smalltalk, from where the block syntax came) are at their heart imperative OO languages. So `do` is natural. Most closures passed in Ruby or Smalltalk would have side effects. This is no longer true in Scala.

We should also not invent many different constructs that mean the same thing.

So, one valid choice would still be: Do nothing. If you want an argument that is a multi-line lambda, use braces, or arrange the lambda vertically, as I had initially proposed. Sure, nobody does it like this now because it costs an extra line. But it might actually lead to clearer code.

If we _must_ invent a parens killing operator, I think `with` is a reasonable choice, after all.

```scala
xs.map with x =>
  val y = f(x)
  g(y)

xs.collect with Some(n) => n

xs.foreach with i => println(s"next: $i")
```
My main concern about doing this is how to restrict it. All the examples above take lambdas, and that looks OK. But what about
```scala
xs.map with f
```
? You should be able to eta-reduce a lambda `x => f(x)` to just `f` without changing its context, so that would imply that this should be legal. But then, don't we also have to accept
```
sqrt with x + x
```
? People will write that sort of code to save a pair of parens! But that's where I think we have made matters worse, not better. 

One possible choice would be to restrict the type of the right operand of `with` to some function type (or maybe accept call-by-name parameters as well?) That's a bit weird, since there is no syntactic reason for this restriction. But it might be a workable compromise.

**Comment by odersky** [comment]
One idea which might be attractive is to restrict `with`'s right operand to call by name arguments or lambdas. That means, `with` is a visual cue that its argument is not strict. `with` introduces significant indentation by itself, so the following works:
```scala
val z = 
  optInt.fold with
    println("nope")
  with x => 
    val y = f(x)
    println(y)
```
The indentation is prompted in one case by `with` and in the other by `=>`.

People have often wished that `{...}` would indicate by-name arguments only. That did not work since we also want sometimes multi-statement by-value arguments, and `{...}` was the only way to get them. But with significant indentation we _could_ introduce two operators that distinguish the two cases. 

If we do this, another question is whether `:` should then apply only to by-value arguments, or whether we want to keep it general.

**Comment by lihaoyi** [comment]
`with` suffers the same problem as `case`: it currently has a very specific meaning (mixin traits) and it is the wrong meaning for what we have here. It is also a very long keyword for what in my experience is a very common operation.

Here's two more ideas:

1. How hard would it be to make do without a keyword? Could we use some combination of lenient parsing + post-validation to allow syntax like this?

```scala
val foo = bar
  .map x => 
    val y = x + 1
    y + 1
  .foreach x =>
    val y = x + 1
    println(y)
```

It seems we would need up to 1 line of lookahead in the lexer/parser, which seems like something that can be afforded. I haven't thought through all possible ambiguities, but it seems to me like with some fiddling this could work. We already do a lenient-parse+post-validation step in parsing lambda argument lists anyway.

2. Could we introduce a new keyword? F# uses the `fn` keyword, which fits perfectly here:

```scala
val foo = bar
  .map fn x => 
    val y = x + 1
    y + 1
  .foreach fn x =>
    val y = x + 1
    println(y)
```

Short, unambiguous, and precisely meaningful. Sure it would be introducing a new keyword, but I think for such a common operation it is worth is v.s. overloading an existing keyword that doesn't really fit

**Comment by odersky** [comment]
I did some exploration, looking at actual usages of `{ ... =>` in out codebase. That made me more skeptical about `with`.  For instance, pairing with `exists` is really bad:
```scala
  sym.baseClasses.exists with ancestor =>
    ancestor.hasAnnotation(jsdefn.EnableReflectiveInstantiationAnnot)
```
This is a LOT less clear than the original
```scala
  sym.baseClasses.exists { ancestor =>
    ancestor.hasAnnotation(jsdefn.EnableReflectiveInstantiationAnnot)
  }
```
(and `do` would be just a bad in its place). This shows again that no single keyword can express function application. The only thing one could do is have a keyword that marks the _start_ of a function and that leaves application silent. `case` was one example of this, `fn` would be another. But it's still awkward to a degree where I would prefer we leave it in braces.

Can we use just `:`? Only if we severely restrict its use since otherwise we would get ambiguities with type ascription. E.g.
```scala
  sym.baseClasses.exists: ancestor =>
    ancestor.hasAnnotation(jsdefn.EnableReflectiveInstantiationAnnot)
```
would theoretically work since the type in a type ascriptions cannot be a naked function type without parens around it. But the longer the parameter list gets the harder it would be to parse (for humans).

**Comment by lihaoyi** [comment]
If `:` is unambiguous (assuming some amount of leniency/lookahead to see the `=>` at end-of-line) then `sym.baseClasses.exists: ancestor =>` seems to me like the best option so far.

It wouldn't be the first time `:` and `=>` are parse differently depending on enclosing context and newlines/semicolon-insertion:

```scala
@ object foo{
    identity: Int => Int
    println("this is a self-type followed by a expression " + identity)
  }
defined object foo

@ foo
this is a self-type followed by a expression ammonite.$sess.cmd9$foo$@555856fa
res10: foo.type = ammonite.$sess.cmd9$foo$@555856fa

@ object foo{
    (identity: Int => Int)
    println("this is a type ascription " + identity)
  }
cmd11.sc:3: missing argument list for method identity in object Predef
  println("but this one is " + identity)

@ object foo{
    val x = {identity: Int => Int}
    println("this is a bound lambda argument header followed by an expression " + x(1))
  }
defined object foo

@ foo
this is a bound lambda argument header followed by an expression object scala.Int
```

While not ideal, it seems to have caused little enough unhappiness in the past, so I wouldn't mind pushing the envelope a little bit, and overloading `:` for opening indented blocks as well

**Comment by Ichoran** [comment]
I appreciate the effort, but don't think the results from any proposal are particularly visually pleasing or sufficiently general.  I'm somewhat puzzled about why people aren't more concerned about the generality/simplicity.

I think we should have a good story about how to select when and where to put a brace-killer.  I don't think memorizing a dozen or more cases is a good story.  This suggests to me that the solution is that brace-killing happens always.

Maybe you require no syntax:

```
2 + 5 *
  3 + 8
```

Or maybe you do:

```
2 + 5 * :
  3 + 8

2 + 5 * ...
  3 + 8

2 + 5 * do
  3 + 8
```

but I don't think it works to have to pick and choose the cases.  I think _even_ `if/then` and `for/yield` are too much to remember if we're going for simplicity.

Python doesn't make you remember a pile of cases. From a random example on the web:

```
    temperature = float(input('What is the temperature? '))
    if temperature > 70:
        print('Wear shorts.')
    else:
        print('Wear long pants.')
    print('Get some exercise outside.')
```

Do you see that `else:`?  The `:` is totally unnecessary because of the keyword `else`.  And yet `:` provides consistency in introducing a new block indent.

Yes, it's just (redundant) punctuation.  But punctuation is important:

> do you see that else: the : is totally unnecessary because of the keyword else and yet : is required for consistency yes its just punctuation but punctuation is important

I think the discussion about `case` vs `do` vs `with` is ultimately much less important than allowing generality and avoiding context-sensitivity.

**Comment by odersky** [comment]
@ichoran Putting `:` behind `else` might work well but is out-of-scope for Scala since it would create a different dialect. I believe all we _might be_ reasonably able to do is make braces optional. But we cannot require `:` where none was required before.

**Comment by krakel** [comment]
little joke
I like this Pascal like syntax:
  `... : ... end`
It looks like:
  `... begin ... end` // we should add this as new keyword
Better than:
  `... { ... }`
end joke

These : used as begin of a indentation looks like an alias of a { without a corresponding a }. You can show this with a simple \\{ or {{ at the end of the line. The : should be separated with a space (xyz : not xyz:) for better reading.

**Comment by Ichoran** [comment]
@odersky - I don't understand what you mean--surely having a brace-free style is making a different dialect?  (Just like `xs take 2` is a different dialect than `xs.take(2)`?)

Anyway, we would only require `:` to open a brace-free block, but we'd _always_ require it to open a brace-free block.

In 2.12/13, `else` does not begin a block, so it would continue to work if it had to be `else:` when followed by a block.  (One could argue that `else:` vs `else` is too subtle to notice, but the compiler could also insist on non-confusing indentation following `else` so that you couldn't make it look like a block with `else` and still compile.)

**Comment by krakel** [comment]
I should describe my idea better.
For normal brace-free style you do not need a block opener. In the rare case where you need a block exceptionally you have to use a block opener. With the brace style we use a single { with block closer at the end. In the brace-free style, we need another block opener. The choice here uses the : . 

The first time I saw that, I was a bit confused. The other suggestions, like do or with, do not feel better either. That's why I wanted to propose a block opener for the brace-free style, which is similar to the normal style. My suggestion here is the double {{ . (I know { on a brace-free style, but we also use () )

Both the single { and the double {{ carry the same message, start a new block, but the latter with brace-free style. The double {{ does not collide with other constructs either.
Only when a user starts a block in a block in brace style should the compiler issue a warning (use spaces between the braces) .

**Comment by aesteve** [comment]
For partial application, I guess some languages (Elm?) do use `\`

```
elements.foreach \elem =>
    doSomeStuffWith(elem)
    println(elem)

elements.foldLeft(0) \elem =>
    doSomeStuffWith(elem)
    println(elem)
```

I did get use to it, some I'm obvisouly biased, but a bit curious if it could be a potential candidate.

**Comment by jxtps** [comment]
Just to clarify for us regular users who are not compiler / language experts, the issue with #2 is the ambiguity as to precisely what part of the indented code is meant for each part of the multi-parenthesis function? I.e.
```scala
def foo(i:Int)(j:Int):Int = i+j

foo
  println("Hello")
  1
  println("World")
  2
  3
```
is completely ambiguous, whereas: 
```scala
foo {
    println("Hello")
    1
  } {
    println("World")
    2
    3
  }
```
is fully specified. 

If this is indeed the issue, how about adding another level of indentation for when you go to the next set of parentheses, like so: 

```scala
foo
  println("Hello")
  1
    println("World")
    2
    3
```

Or did #2 refer to something else?

**Comment by jxtps** [comment]
For #1, definitely go with Martin's initial second suggestion, "Make indentation significant after class, object, etc headers without requiring a semicolon. This has the problem that it is not immediately clear whether we define an empty template or not. E.g. in

```scala
object X extends Z

 // ...
 // ...

 object Y
```
it is hard to see whether the second object is on the same level as the first or subordinate."

Well that's a funny rejection of indentation based syntax! ;) 

I'd argue that there's no delimiter that will make the "is object Y subordinate or not" all that clear when there's vertical distance between X and Y. 

In fact, I haven't seen any substantive argument against this option, so we might as well go with the lightest syntax possible.

**Comment by odersky** [comment]
To clarify what I mean when I say we do not want to create a dialect: It requires that _code that is currently well-indented means exactly the same with optional braces and without_. 

That rules out completely silent syntax for objects and classes.

**Comment by jxtps** [comment]
So to clarify, the concern is that
```scala
class C
  def foo = 1
```
could be interpreted as either: 
```scala
class C
def foo = 1 // <- top level method, because these are allowed in dotty?
```
or
```scala
class C {
  def foo = 1
}
```

So the ambiguity arises from the possibility that a sloppily indented file written in the braces era would mean something different in the indentation era? 

Based on "code that is currently well-indented means exactly the same with optional braces and without" my interpretation must be incorrect, so there is some other ambiguity that I don't see?

**Comment by krakel** [comment]
@jxtps 
> ```scala
> foo {
>   println("Hello")
>    1
>  } {
>    println("World")
>    2
>    3
>  }
> ```
your code should be written in brace-less style:
```scala
foo
   println("Hello")
   1
:   // <- start a new brace-less block
   println("World")
   2
   3
```

**Comment by jxtps** [comment]
@krakel 
Right, and that extra `:` looks... terrible? Hence my suggestion for an additional level of indentation instead. 

It feels reasonably natural in the world of indentation: a new level of indentation means a new block, which is the same as putting braces around that part of the code, which is what we were looking for here.

**Comment by Ichoran** [comment]
@jxtps - `bippy{a}{b}` is not the same as `bippy{a{b}}`, and the latter needs to be easily encoded, so your idea won't work.

**Comment by jxtps** [comment]
Excellent point @Ichoran, thank you! The syntax I suggested would be more suitable for the `bippy{a{b}}` case. For the `bippy{a}{b}` case we need to allow ending an indentation block only to start a new one. We already have an `end` token. Reuse it like so: 
```scala
foo
  println("Hello")
  1
end
  println("World")
  2
  3
```

**Comment by odersky** [comment]
@jxtps Yes, it comes down to what we regard as properly indented. So my comment above was misleading. I have edited it now to make it clearer.

**Comment by jxtps** [comment]
"**Braces optional**" is an interesting way of thinking of the problem. What if we also think of it from the mirrored perspective: "**indentation means braces**", i.e. wherever there is (single level) indentation, imagine there are braces around it - no new keywords or delimiters (`:`, `...`, `do`, `where`, ...). What happens then, and what would need to change to make that work?

Let's consider two rules: 
1. Single indentation means braces. 
1. Double+ indentation means continuation of the last line (like ending the previous line with `\` in e.g. bash) - the intention here is to meaningfully line up the next line's code with a suitable column of the preceding line. 

If we look at the meat of https://github.com/lampepfl/dotty/pull/7024#issuecomment-528773997 (yes I dropped the new `then`, `do` etc since they're superfluous for multi-line cases): 

```scala
if  x < 0 && 
    y > 1  // <-- extra indent means continuation of last line = it's "inside the ()"
  -x 
else 
  x

while x >= 0 && 
      y > 1 // <-- extra indent means continuation of last line = it's "inside the ()"
  x = f(x)

for x <- xs 
    if x > 0 // <-- extra indent means continuation of last line = it's "inside the ()"
yield 
  x * x

for x <- xs
    y <- ys // <-- extra indent means continuation of last line = it's "inside the ()"
  println(x + y)
```
Since the "signature" of the control flow statements is roughly `while(cond) { block }` then we can see how this naturally translates to library code of similar signatures. 

Something along these lines avoids excess keywords/tokens and is very "airy", but there's of course a concern with distinguishing between two levels of indentation: will users grok and realize the difference? And it probably violates the "no new dialect" doctrine.  

We'd need to figure out at least how it would work in these additional cases (some of which have been covered above): 
```scala
// General function invocations, with simple arguments f(1,2), blocks f({a},{b}) & blocks taking arguments f(x => ...): 
f(x, y)
f(x => a, y)
f(x => a, y => b)
f(x)(y)
f(x => a)(y)
f(x => a)(y => b)
f(g(x))
f(x => g(y))
f(x => g(y => b))

// When code overflows one line: 
val myReallyLongVariableName: MyReallyLongClassName = MyWayTooLongInitializer0 * 
  MyWayTooLongInitializer1 // How far do we indent? Min/max? What's clear?

// When chaining method calls (which is a version of code overflowing a line): 
xs
  .map(_.foo()) // How far do we indent? Min/max? What's clear?
  .flatMap(_.bar())
  ...

// When a method takes a lot of parameters, for clarity (also code overflowing a line)
xs.foo(
  a = 10, // How far do we indent? Min/max? What's clear?
  b = false, 
  ...
)

// There are also some code conversion concerns: 
val a = x * 
  y + z // With "old school indent" this risks being interpreted as x * { y + z } instead of x * y + z!
```

Maybe I'm off in space here (no pun intended, really), but on some level this entire indentation thing is too, so yeah, it would be great if we could get the benefit of braces optional without too much additional keyword / token baggage. And you can't get fewer tokens than 0 (not counting whitespace ;)

**Comment by odersky** [comment]
I have a hard time believing that "indentation means braces" would work without creating a dialect. 

I propose the following procedure instead:

 1. Decide on rules to check that today's Scala code with braces is properly indented
 2. Implement these rules in the compiler or a format checker.
 3. Identify places where braces would be redundant for well-indented Scala programs. I.e indentation
     and braces always agree.
 4. Allow to drop the braces in these situations.

That way we are sure not to create a dialect: Some braces are simply optional for correct Scala programs, the same way some braces are optional now.

**Comment by odersky** [comment]
I have implemented in #7185 a version that does not require a `:` after a class or object signature. I believe it is overall an improvement. It does require we tighten rules for well-indented programs. In particular:

 1. All definitions inside a class or similar have to be indented relative to the start of the class.
 2. The definitions following a class without a body _cannot_ be indented relative to the start of the class.

I have seen some reasonable code violating the second requirement. For instance:
```
trait A
  case class B() extends A
  case class C() extends A
```
This would have to be outlawed since otherwise the code above would be regarded as
```
trait A {
  case class B() extends A
  case class C() extends A
}
```
So we would have to make at least one controversial restriction on what is well-formed indentation to make optional braces behind class signatures work. Still, I think it is worth doing that.

**Comment by jxtps** [comment]
If we have to have a token to tell when we go brace-less in a method invocation (which it seems like we have to), then there doesn't seem to be an ideal candidate, but `:` seems to be the least bad option, with the addition of allowing `xs.foreach: x =>` to start <indent> like @lihaoyi [suggested](https://github.com/lampepfl/dotty/issues/7136#issuecomment-527037434). 

There's lots of weirdness in the syntax when you start enumerating all the combinations of functional forms and parameter styles. This list still needs to be worked through: 
```scala
f(x, y)
f(x => a, y)
f(x, y => b)
f(x => a, y => b)
f(x)(y)
f(x => a)(y)
f(x)(y => b)
f(x => a)(y => b)
f(g(x))
f(x => g(y))
f(x => g(y => b))
```

**Comment by Ichoran** [comment]
@odersky - Keep in mind that if a block is, in isolation, well-indented and braces are redundant, it does not follow that this remains true after braces are removed from neighbors.  For instance,

```scala
foo{
  println("parameter block 1")
  p
} {
  println("parameter block 2")
  q
}
```

meets all the criteria for both blocks, as far as I can tell (modulo some choice about how `} {` should appear--one line or two, etc.).  And each single removal works:

```scala
foo {
  println("parameter block 1")
  p
}
  println("parameter block 2")
  q

foo
  println("parameter block 1"
  p
{
  println("parameter block 2")
  q
}
```

But obviously both together is a problem:

```scala
foo
  println("parameter block 1")
  p
  println("parameter block 2")
  q
```

**Comment by jxtps** [comment]
@Ichoran 
I thought we'd have to have `:` to start a braceless block for most method invocations - otherwise you also have problems like this: 
```scala
// a = (x * y) + z in scala2
val a = x *
   y + z

// but interpreted as a = x * (y + z) in scala3
val a = x * {
  y + z
}
```
Still not clear on how to write the invocation you describe in braceless style - none of the forms I've seen jump out as being all that great:
```scala
foo: 
  println
  p
: // Variations: ":", "end", "end:" - all have issues. 
  println
  q
```

**Comment by flavomulti** [comment]
How do we handle fold when there are 2 parameter lists?
```scala
val value = Option(12)
val x = value.fold(10)(v => v+1)
```
I'm not sure what the approach is, I've not had anything work. For this trivial case I would just keep it on one line of course
```scala
val x = value.fold:
  10
???
```

**Comment by Ichoran** [comment]
Well, I think the conclusion is that only the last block can be brace-optional.

```scala
xs.foldRight(0)
  (acc, x) => acc + x
```

and that explicit implicit parameters require braces (using a 2.12 example):

```
xs.map{ p =>
  val q = foo(p)
  baz(q, bar(q))
}(cbf)
```

**Comment by aappddeevv** [comment]
While I am not against it, it feels inconsistent. Perhaps there are some more advanced rules around to allow braces optional to still apply in this senario.

**Comment by lihaoyi** [comment]
> Well, I think the conclusion is that only the last block can be brace-optional.

Using a keyword to delimit blocks allows multi-param-list brace-free function calls to look OK in my opinion:

```scala
val sum = xs.foldLeft fn
  0
fn (acc, x) => 
  acc + x

xs.map fn p =>
  val q = foo(p)
  baz(q, bar(q))
fn
  cbf
```

```scala
val sum = xs.foldLeft do
  0
do (acc, x) => 
  acc + x

xs.map do p =>
  val q = foo(p)
  baz(q, bar(q))
do
  cbf
```

This would be a trivial desugaring to `{}`s:


```scala
val sum = xs.foldLeft {
  0
} { (acc, x) => 
  acc + x
}

xs.map { p =>
  val q = foo(p)
  baz(q, bar(q))
} {
  cbf
}
```

Using `:` only is problematic because `:` alone on its own line looks weird:

```scala
val sum = xs.foldLeft :
  0
: (acc, x) => 
  acc + x

xs.map : p =>
  val q = foo(p)
  baz(q, bar(q))
:
  cbf
```

Making these multi-arg-list function calls pretty is one of the reasons I think using a two-character keyword like `do` or `fn` is superior to using an operator like `:`, even though neither `do` nor `fn`'s english meanings are 100% on target

**Comment by aappddeevv** [comment]
I could get used to the last sum with : I think.  But not every 2nd Param list is a function and it would look awkward if it is just a short expression in which case why use indent syntax. I’m beginning to think that this syntax is good at getting rid of obviously redundant braces and braces for blocks with more than one statement.

**Comment by Ichoran** [comment]
@lihaoyi - I was trying to adhere to Martin's new "strict elision" policy--that is they are _optional braces_ with no extra syntax, to make it as little of a language split as possible.

If the feature is _merely_ optional braces, nothing else, then only the last block could work since the syntax is otherwise ambiguous.

**Comment by lihaoyi** [comment]
In my opinion we have to go "all the way" and handle all common uses cases of braces with indentation if we want this to be usable. Although I prefer python-style brace-free code over java-style braces-everywhere code, I think a mixed braces-in-some-places-but-not-others code would be the worst scenario of all: neither visually coherent, nor easy for people to remember the rules for, nor similar to any other language out there.

What language out there has indentation-delimited blocks but still requires heavy use of brace-delimited blocks on a day-to-day basis? There isn't one.

I wouldn't be exaggerating saying braces-for-function-calls-but-not-for-classes would make it the worst surface syntax of any programming language I've ever seen.

**Comment by LPTK** [comment]
Wouldn't a simple "application" special operator solve the remaining cases?
It preserves the simplicity of the base rules (can elide braces after special operators like `=>`, `=` and  `then`)

**Comment by eed3si9n** [comment]
@LPTK since `@ (acc, v) =>` looks similar to a pattern match, [`<|`](https://github.com/lampepfl/dotty/issues/7136#issuecomment-526698122) might look better?

```scala
ls.fold <|
   val x = foo()
   (x + 1, x - 1)
<| (acc, v) =>
   println(v)
   (acc + v._2, v._1)
```

**Comment by AndreVanDelft** [comment]
### A different perspective: 

Allow for a Smalltalk-like method call syntax and combine that with indentation-implied-braces.

### Example
```scala
def unless_do(test: => Boolean, code: => Any) = if test then code

val v = unless: test, do: w
```

### Rules

* The colon should immediately follow the partial method name
* The complete method name is constructed by concatenating the partial names with an underscore in between
* The partial names may be keywords; i.e. in an expression, an identifier immediately followed by a colon is never classified as a keyword
* Varargs are possible; they don’t need to be the last arguments in the call and there may even be multiple varargs groups

### More examples
```scala
do: code,
while: test

val v = do: code
      , while: test

for i <- from: 0, to: n do something
for i <- from: 0, until: n, step: 2 yield i

val v = if:     test1 -> v1
      , elseIf: test2 -> v2
      , elseIf: test3 -> v3
      , else:            v0

xs.foreach: i => println: s”next: $i"

val foo = ( bar.map: x => 
              val y = x + 1
              y + 1
          ).foreach x =>
              val y = x + 1
              println: y
```
In last example the method-chaining still requires a pair of parentheses;
The version that Haoyi proposed earlier in this thread also has parentheses, namely for the call to `println`.
### Notes

This proposal does not distinguish between by name arguments and others.

The new syntax would clash with the current Scala syntax for typed expressions: `PostfixExpr ':' CompoundType`. AFAIK the typed expression syntax is rarely used. It might be replaced by something else, such as `expr.as[CompoundType]`.

I am not yet sure whether there would be other problems with this proposal; I may have overlooked a few things.

In the example with calls to `map` and `foreach`, the braceless lambda parameters start with `x =>`; hopefully this syntax is feasible.

The method definition could have a Smalltalk-like syntax:
```scala
def unless:(test: => Boolean), do:(code: => Any): Unit = if test then code
```
This will be harder to parse since after the first colon in the current syntax the return type specification is expected. Leaving out the colons after the partial names could solve this, but the parser should there allow for identifiers that are normally keywords.

### Background

Parentheses are about as noisy as braces in code. In Scala they are currently used for:

* expressions with binary operators
* method calls
* method definitions
* tuples
* `for`, `if` and `while` statements

Parentheses after `for`, `if` and `while` are already disappearing.
For their use in method calls the Smalltalk style, augmented with separating commas, offers a good alternative.

In [SubScript](http://subscript-lang.org), an extension to Scala, the Smalltalk-like call syntax works.

See also my presentation [*Broaden Scala* at LambdaDays 2018](https://www.lambdadays.org/lambdadays2018/andre-van-delft). Video and slides are available.

**Comment by odersky** [comment]
@lihaoyi 

> What language out there has indentation-delimited blocks but still requires heavy use of brace-delimited blocks on a day-to-day basis? There isn't one: not python, not f#, not haskell, not coffeescript

In fact, it would be good to catalogue how these languages treat multi-statement function arguments. That could give us some ideas what to do.

**Comment by AndreVanDelft** [comment]
An addition to my previous posting:

Smalltalk style parameter lists would diminish the need for named parameters and for infix method calls.

E.g. instead of `0 to 10` you would write `from: 0, to: 10`. A bit longer, but in my view clearer because it symmetrically describes the role of both the lower bound and the upper bound of the interval.

`0 to 10 by 2` looks good, but that is means of a hack: the `10 by 2` part creates a tuple, which is passed to an overloaded version of `to`. IMO those are bad signs.
The Smalltalk style alternative would be `from: 0, to: 10, by: 2`.

The whitespace operator for infix calls could be downgraded to become nonstandard, limited to scopes with an `import language` feature statement. Then would become available for other uses.

In my presentation at LambdaDays, I argues for opening up the scanner and parser of the Scala compiler so that experimentation and dialects become easy. This would widen the market for Scala; it could become the basis of many DSLs, e.g. for parsing, statistics and simulation. 

In these DSLSs whitespace could be given a different purpose in expressions than infix notation.
E.g. for grammar description whitespace would denote sequence.

**Comment by LPTK** [comment]
@eed3si9n I agree that `<|` is the most natural operator name for this semantics.

The problem is that `<|` looks too much like a _normal_ operator, and people are probably already using it as such!

On the other hand, `@` is currently **not** a valid operator name. It's a special symbol in Scala which currently only makes senses in annotations and patterns. So it would make more sense a a _special_ operator that has super powers (smallest precedence; can introduce newlines).

**Comment by odersky** [comment]
> In fact, it would be good to catalogue how these languages treat multi-statement function arguments. That could give us some ideas what to do.

Here's a partial answer from what I could find browsing the web. Please correct where I am wrong or missing things:

 - Python:  No multi-statement arguments are allowed
 - F#: No special syntax, use parentheses
 - Haskell: No special syntax, use parentheses
 - Coffeescript: Indentation is significant for the last argument, parentheses are required for preceding arguments, i.e similar to what @ichoran brought up earlier. However, [people recommend to put all arguments in parentheses](https://stackoverflow.com/questions/7532616/multiline-function-calls-in-coffeescript).

From these comparisons it does not follow to me that we necessarily have to do something. Keeping braces looks like a viable option. 

Assuming for the moment we want to do something, I have assembled possible options with the same benchmark program:

**Using braces:**
```scala
logging("stdout") {
  xs.map {
    case a => ???
    case b => ???
  }
  .filter { x =>
    x >= 0
    && x < limit
  }
  .exists { x =>
    val y = x * x
    y >= limit
  }
  .fold(zero) { (x, y) =>
    val z = x + y
    z * z
  }
}
```
**Using `fn`**
```scala
logging("stdout") fn
  xs.map fn
    case a => ???
    case b => ???
  .filter fn x =>
    x >= 0
    && x < limit
  .exists fn x =>
    val y = x * x
    y >= limit
  .fold(zero) fn (x, y) =>
    val z = x + y
    z * z
```

**Using `<|`**
```scala
logging("stdout") <|
  xs.map <|
    case a => ???
    case b => ???
  .filter <| x =>
    x >= 0
    && x < limit
  .exists <| x =>
    val y = x * x
    y >= limit
  .fold(zero) <| (x, y) =>
    val z = x + y
    z * z
```

**Using `:`**
```scala
logging("stdout"):
  xs.map:
    case a => ???
    case b => ???
  .filter: x =>
    x >= 0
    && x < limit
  .exists: x =>
    val y = x * x
    y >= limit
  .fold(zero): (x, y) =>
    val z = x + y
    z * z
```

**Using `@`**
```scala
logging("stdout") @
  xs.map @
    case a => ???
    case b => ???
  .filter @ x =>
    x >= 0
    && x < limit
  .exists @ x =>
    val y = x * x
    y >= limit
  .fold(zero) @ (x, y) =>
    val z = x + y
    z * z
```

**Comment by odersky** [comment]
I have now started to use the new syntax heavily when working with the compiler. Here are some first impressions from ~3 days of intensive usage of working quiet control syntax and optional braces (but not using `:` at EOL).

 - I had a surprisingly hard time getting used to the `if-then-else` syntax. My finger memory wants to type `(` after `if`. I expect this to be transitory, however. 
 - Reading code with the quiet control syntax is significantly easier than reading code with `(...)` conditions, even though I am used to the latter style from 25 years experience.
 - Not having to write braces is generally a god send. It makes you realize in how many ways braces interfered before with your workflow.
 - If braces are omitted, code gets squashed vertically (at least 10% fewer lines in my experience). Sometimes this requires a conscious decision to compensate by inserting empty lines or end markers. Once that is done, the resulting code is easier to read than code with braces.

**Comment by AndreVanDelft** [comment]
#### Using Smalltalk-style method calls:
```scala
logging("stdout"):
  ((( xs.map:
    case a => ???
    case b => ???
  ).filter: x =>
    x >= 0
    && x < limit
  ).exists: x =>
    val y = x * x
    y >= limit
  ).fold(zero): (x, y) =>
    val z = x + y
    z * z
```
I am not really happy with that. An extra rule for dot concatenation could help, maybe. I am thinking of this:
```scala
term
  indented
  ...
.selector
  ...
```
would be shorthand for
 ```scala
(term
  indented
  ...
).selector
  ...
```
Then the benchmark result would become what Martin wrote under * **Using `:`** *

**Comment by lihaoyi-databricks** [comment]
> Please correct where I am wrong or missing things

I believe F# sometimes uses parentheses and sometimes doesn't: F# -> F# method calls usually are curried Ocaml-style without parentheses `f a b c`, while F# -> C# method calls require parentheses `f(a, b, c)`.

Having used F# for a reasonable amount of time (O(10k) lines written, over maybe a year and a half) I found the mix of parens-vs-no-parens a big headache I never quite got used to. It's an inconsistency that permeates all your code, and even after a year+ I always had to spend time thinking which way a particular function had to be called.

Coffeescript also allows you to elide parentheses, and as well as commas, in method call argument lists. But having maintained a large (500k line) codebase in it, our experience was that it was a very bad idea. When both multi-expression argument lists and multi-statement blocks both are delimiter free, and any expression can be a statement (i.e. similar to Scala) it becomes incredibly difficult to figure out whether a bunch of code is a bunch of parameters passed to a single function, or a bunch of statements in a block.

In contrast, languages like Python with no-curlies-but-required-parentheses make the distinction immediately obvious, probably no more difficult to parse visually than languages like Javascript that require both

**Comment by AndreVanDelft** [comment]
BTW The last part
```scala
.fold(zero): (x, y) =>
    val z = x + y
    z * z
```
would become in Smalltalk style:
```scala
.inject: zero, into: (x, y) =>
    val z = x + y
    z * z
```

**Comment by odersky** [comment]
@AndreVanDelft I believe Smalltalk like syntax is intriguing, but don't see a way to do this for 3.0. Maybe later...

**Comment by aappddeevv** [comment]
I’m using it exclusively now for my one project. I would not want to go back to braces even if it is not perfect. For me, there was no jarring change, no steep learning curve and no readability issues for crazy hacked code I need to refactor or 2.x code I need to review for older libs. But I’m not a full time programmer, I just jump in and out of it.

**Comment by krakel** [comment]
I would like to advertise my proposal again. Move the closing bracket to the beginning.

**Using `{{`**
 
 ```scala
 logging("stdout") {{
   xs.map {{
     case a => ???
     case b => ???
   .filter {{ x =>
     x >= 0
     && x < limit
   .exists {{ x =>
     val y = x * x
     y >= limit
   .fold(zero) {{ (x, y) =>
     val z = x + y
     z * z
 ```

**Comment by Ichoran** [comment]
@krakel - I like my parentheses, brackets, and braces balanced.  (Angled brackets too, in languages that use them as brackets.)  I suspect there are enough people with a visceral aversion to imbalanced parens etc. that this is not feasible.

@AndreVanDelft - Memorizing a bunch of superfluous method argument names isn't fun.  I would reject anything that requires it.

@odersky - Did you still want to have optional braces?  If yes, you have to consider how all of these things work with braces _plus_ the block opener:

```scala
logging("stdout") fn {
  xs.map fn {
    case a => ???
    case b => ???
  }.filter fn { x =>
    x >= 0
    && x < limit
  }.exists fn { x =>
    val y = x * x
    y >= limit
  }.fold(zero) fn { (x, y) =>
    val z = x + y
    z * z
  }
}

logging("stdout") <| {
  xs.map <| {
    case a => ???
    case b => ???
  }.filter <| { x =>
    x >= 0
    && x < limit
  }.exists <| { x =>
    val y = x * x
    y >= limit
  }.fold(zero) <| { (x, y) =>
    val z = x + y
    z * z
  }
}

logging("stdout"): {
  xs.map: {
    case a => ???
    case b => ???
  }.filter: { x =>
    x >= 0
    && x < limit
  .} exists: { x =>
    val y = x * x
    y >= limit
  }.fold(zero): { (x, y) =>
    val z = x + y
    z * z
  }
}

logging("stdout") @ {
  xs.map @ {
    case a => ???
    case b => ???
  }.filter @ { x =>
    x >= 0
    && x < limit
  }.exists @ { x =>
    val y = x * x
    y >= limit
  }.fold(zero) @ { (x, y) =>
    val z = x + y
    z * z
  }
}
```

None of these look very good to me, but to my eye `@` is now the least bad.  Also, none of these match how I'd have written the original with braces:

```scala
logging("stdout"){
  xs.
    map{
      case a => ???
      case b => ???
    }.
    filter{ x =>
      x >= 0 &&
      x < limit
    }.
    exists{ x =>
      val y = x * x
      y >= limit
    }.
    fold(zero){ (x, y) =>
      val z = x + y
      z * z
    }
}
```

I find that with this formatting, the visual weight of the braces is greatly reduced.  By using dots at end of lines instead of the beginning, the method names leap out even more, making the control flow easier to manage.  But with brace-elision, the dots are absolutely required at the beginning or the meaning changes.

So I'm not entirely convinced the advantages of going brace-free are quite as big as suggested.  I think better style choices can help with existing syntax.

**Comment by AndreVanDelft** [comment]
Martin:
> I believe Smalltalk like syntax is intriguing, but don't see a way to do this for 3.0. Maybe later..."

I understand.

More layout exploitations are possible. I gave some 'Wild Ideas' in my LambdaConf talk [PDF](https://www.lambdadays.org/static/upload/media/1519724358629669andrevandelftbroadenscala.pdf).

Some of these exploitations will conflict with others. I'd propose to make consistent dialects available through `import language`, in some version after 3.0. Developers can then experiment and select the fittest dialects for their tasks. The most popular dialect could then even become the main language. But IMO that should only happen after years of ideas and experience; with an exception for the layout-implied-braces.

**Comment by lbkb** [comment]
The problem noted by @jxtps with
```scala
val a = x *
   y + z
```
only affects lines ending with a binary operator, right? So we could save existing behaviour by adding a rule like: "Indentation after a line ending with a binary operator is treated as line continuation".

That looks natural to me:

```scala
if a < b &&
   b < c
  1
else 
  2
```
.. provided lines 2 and 3 are indented differently.

**Comment by lbkb** [comment]
Then with ` : ` as parens-killer you end up with something pretty light: 

```scala
logging("stdout")
  xs.map 
    case 0 => 1
    case x => x
  .filter : x =>
    x >= 0
    && x < limit
  .fold 
    zero
  : (x, y) =>
    val z = x + y
    z * z
```
Note that a space before `:` makes it look more operator-like.

**Comment by odersky** [comment]
The merged PRs #7185 and #7235 implement the following scheme:

Braces are optional for correctly indented code

 - after `=`, `=>`, `<-`, 
 - in parts of control structures, i.e. if-then-else, while-do, try-catch-finally, match, and for-yield/do
 - after headers of classes, etc, where a list of definitions is expected.

A `:` at end of lines is significant only if the option`Yindent-colon` is given. 

This means that we defer until later the question  whether to adopt `:`, or use something else instead, or just keep on using parentheses or braces.

**Comment by odersky** [comment]
After having worked with the new rules (without `:`) for a while, I observe that omitting braces after class or object headers feels different from omitting braces everywhere else. Everywhere else (i.e. after `=`, `=>`, `<-`, `then`, `else`, `do`, `yield`, `try`, `catch`, `match`, and so on) omitting the braces feels completely natural and improves legibility. The reason is that in these places _something has to follow_. I.e. if I write
```scala
def foo =
  abc
```
The `abc` is mandatory, it cannot be left out. The reader parses visually the indented block as the continuation, and it does not matter whether that block consists of one or more statements. For someone parsing the outline of code, the two snippets
```scala
def foo =
  println("abc" ++
    "xyz")
```
and
```scala
def foo =
  println("abc")
  "xyz"
```
are equally clear and braces are needed in neither case. So, I find making braces optional here is a clear win because it improves writability and readability at the same time.

With class and object headers it's different. Here when I see
```scala
class Rational(x: Int, y: Int)

  /** The numerator */
  def numer = x

  /** The denominator */
  def denom = y
```
I am less sure how to read that. Are `numer` and `denom` definitions in class `Rational` or are they following it? I have to count spaces to decide. The immediate effect was that I felt that in this case having more than two spaces would help. It's also a case where adding a space accidentally would change meaning, which is never the case elsewhere, and which is something we would want to avoid.

By contrast, braces help here. When I see
```scala
class Rational(x: Int, y: Int) { 
...
}     
```
I know that the next definitions `...` are members of `Rational` and I can scan for the closing brace `}` to see where it ends. `end` markers are a superior replacement for closing braces but the problem remains how can I be sure that anything starts after `Rational(x: Int, y: Int)` at all?

Possible solutions:

 1. The minimal viable change: Continue to require braces around class/object/... definitions. This would work in the short term, but might lead to an uncomfortable style in the longer term where programs are a mixture of braces and indentation.

 2. Allow to drop braces after class/object/... only if there is a marker word that indicates that something will follow.  Various choices for the marker have been discussed. I still think that `:` is
far and away superior to all of them. So:
    ```scala
    class Rational(x: Int, y: Int):

      /** The numerator */
      def numer = x
      ...
    end Rational
    ```
 3. Once we have introduced a marker like `:` to replace a block in `{...}` we can also use it for function arguments. That would go back to the originally proposed version.

**Comment by aappddeevv** [comment]
When fiddling with 0.19 I immediately made that mistake on a few class defs without the colon, especially for case classes. It was a bit hard to read without the colon as I was not using an `end` (which I wanted to avoid).

```scala
case class Foo(
  x: Int,
  s: String
): // <-- this colon was very helpful when reading the definition
  def someMethod: String = ???
```

**Comment by arturopala** [comment]
Do you see in the long term brace-less syntax as an alternative or an replacement?

**Comment by jxtps** [comment]
It seems like the primary difference between the class/object case and the method case is less that something *must* follow in the method case, and more that even short classes/objects typically have multiple methods/fields and there's vertical white space between them. 

If we were to consider a method with a longer body:
```scala
def foo(i:Int, y:Int):Int = 
  println("Welcome to foo!")
  
  val a = i * y
  println("A is: " + a)

  for (val b <- 0 until a) print(".")

  println("Goodbye!")
  a+2
```
then it's no longer any more or less clear what's part of the body vs what's part of the enclosing scope compared to classes/objects, right?

> It's also a case where adding a space accidentally would change meaning, which is never the case elsewhere, and which is something we would want to avoid.

Sure, the first line of a method may be protected by "something must follow", but the subsequent lines are not, right?

In my Python experience (with 4 spaces per indent which was the editor's default) there's never been any ambiguity whatsoever about these things. Copy-paste is a little futzy, but I've never found myself in a situation where I wonder what scope a given line of code is in. For the cases where it gets hairy - really long methods / classes - any braces or `end Foo` are typically far away anyway, so they don't seem to solve this problem (to the extent it is one). 

Instead, I would consider going draconian: if you're going to use indentation-based syntax, the indentation shall be 4 spaces per level, no tabs, no exceptions. See [PEP-8](https://www.python.org/dev/peps/pep-0008/#indentation) - I'm guessing the Python folks have hashed this out a lot more over the years than we have ;)

**Comment by odersky** [comment]
> then it's no longer any more or less clear what's part of the body vs what's part of the enclosing scope compared to classes/objects, right?

It does not work that way for me. Visually, the first line after the `def` _must_ be the body since it follows a `=`. Then subsequent lines starting at the same column are clearly part of the same body. We have been trained to read vertical alignment as being in the same list.

But with classes it's different. Here, I have to check whether something is indented relative to the previous line. That's much harder, and suffers from off-by-one-space errors.

Going to 4 spaces might help but it's not something we can require.

**Comment by odersky** [comment]
One downside of using `:` for introducing template definitions is that it works poorly with `given`.

```scala
given c: C:
  def f() = ...
```
When I originally proposed `:`  for introducing template definitions the `given` syntax still used `as` instead of `:`, so that was not a problem then. But I believe we are now settled on the `given c: C` part, so this is now an issue that needs to be resolved. 

The alternative of using `with` does not have that problem, and reads quite well, in fact.

```scala
trait Monoid[T] extends SemiGroup[T] with
  def unit: T

given Monoid[String] with
  def (x: String) combine (y: String): String = x.concat(y)
  def unit: String = ""
```
I am less keen on using `with` for method arguments, for the reasons I have already stated.

**Comment by nafg** [comment]
I really wish all the time being invested in this would be saved for after some more urgent things, for instance getting TASTy to the point where it can be used in Scala 2.x.

**Comment by TheElectronWill** [comment]
> I really wish all the time being invested in this would be saved for after some more urgent things, for instance getting TASTy to the point where it can be used in Scala 2.x.

Definitely agree that it would be good to move forward on TASTy! I don't think we can prevent discussions and experiments around syntax (whatever our opinion), but I'm positive that PRs are welcome :-)

**Comment by bishabosha** [comment]
@nafg @TheElectronWill This is being worked on at the Scala Center: [TASTy Reader for Scala 2](https://github.com/scalacenter/scala/tree/tasty_reader)

**Comment by Ichoran** [comment]
@odersky - I really like the `with` syntax for opening a block for a trait or class, because it is equivalent to mixing in an anonymous trait, which is conceptually equivalent anyway.  That is, the following two should be equivalent in terms of what methods trait C gets; the only difference is whether you can easily name the subset that are not part of A:

```scala
trait A { def a: A }
trait B { def b: B}
trait C extends A with B

trait A { def a: A }
trait C extends A with
  def b: B
```

**Comment by jxtps** [comment]
There is an argument to be made that the braces vs indentation thing should largely be handled by the editor. 

To that end I have filed a feature request with IntelliJ: https://youtrack.jetbrains.net/issue/IDEA-224361 - you may want to consider filing similar requests with your favorite editor, or voting for / promoting such features where applicable / possible.

**Comment by nafg** [comment]
You can change the syntax coloring to make braces a really light color (or
dark depending on your background). And intellij already inserts closing
braces for you. What more can they do exactly?

If you want to go in this direction the broader issue is that sources are
text files. But once TASTy is finalized someone can make a GUI editor for
it that looks like a text editor (it would need to typecheck on the fly) or
an untyped analogue.

**Comment by esarbe** [comment]
Why not regularize the definition syntax for object/classes/traits to match the method definition?

```
class Rational(x: Int, y: Int) =

  /** The numerator */
  def numer = x
  ...


object Foo = 
  ....

trait Bar(x: Int, y: String) =

end trait
```

**Comment by TheElectronWill** [comment]
@esarbe This has already been offered during the discussions on indentation-based syntax. IIRC, the argument against doing this was the following:
- For functions, the left- and right-handside are equivalent/interchangeable.
- But it's not the case for classes nor traits (e.g. two classes with the same methods aren't interchangeable).

**Comment by esarbe** [comment]
@TheElectronWill Ah, interesting. I never considered the `=` for `val`s or `def`s to mean equality but always as assignment, similar to Pascal's `:=`.

**Comment by TheElectronWill** [comment]
Well, `==` still means `equal` and `x = 2` still means `assign 2 to variable x`, but there still is a semantic issue with `class A = something`

**Comment by nafg** [comment]
In a sense, a class is a subtype of its template's structural type. So if
you really want to go down this rabbit hole... how about <:

class Rational(x: Int, y: Int) <:

  /** The numerator */
  def numer = x
  ...

**Comment by mjburgess** [comment]
I'd like to +1 on the strategy of regularising brace-formatted scala code and then allowing braces to be dropped. I think this makes the sales pitch (/impact assessment) less fraught. 

Braces seems somewhat arbitrarily optional at the moment (in the sense that, eg., in Java, every class definition requires them). There's a strong case for saying "always optional", if that can be achieved with minimal cognitive overhead. Its quite common in C to drop braces wherever possible (whether advisable or not) -- I think the appetite for reduced line noise is pretty widespread. 

It seems "with" does make sense for bodies/definitions (,...) as there does feel something similar in intention when saying "with Trait" and "with {body}" -- ie., with Trait (qua mixin) serves as a kind of syntactical abbreviation for the in-place definition. 

Is there a case for the colon  and `with`?

**Comment by odersky** [comment]
A status report:

Current master allows to drop braces around templates after `with`. My initial experience is that it works well. We should wait for a Dotty release cycle where others can try it out before reaching a definite decision on it. 

`:` is currently enabled under -Yindent-colons. It has two uses

 1. as an alternative to `with`, to start a template
 2. to indicate an indented function argument block

Usecase (1) is an alternative to `with` in current master. In the end, we should pick one or the other (or none at all). With the new given syntax - which I assume is settled by now - `:` becomes more awkward than before, e.g.
```
   given intMonoid: Monoid[Int]: 
     ...
```
I find the double use of `:` on a line problematic. `with` reads better in this case:
```
  given intMonoid: Monoid[Int] with 
    ...
```
Usecase (2) is still open. Currently `:` only works at end of line, which rules out idioms like
```
  xs.exists: x =>
    x > 0
```
Enabling this syntax is almost possible. The only clash is with an explicit self type declaration at the start of a class or trait. E.g.
```
  trait T {
    foreach: x =>
      println(x)
```
Here, `foreach: x =>` would be parsed as a self type declaration for trait `T`. There might be ways around that, however: for instance, we could try a different syntax for self type declarations, see #7374.

My current and still provisional tendency is to do nothing about usecase (2) and to retire the -Yindent-colons option for now. That way, we can evaluate the other changes first before taking the next step. I believe a parens killing operator could still be an interesting option at some point, but I see less urgency to put it in Scala 3.0.

**Comment by odersky** [comment]
Closing since it looks there won't be anything standardized for 3.0. We can of course get back to it at some later point.
